components/director/internal/tenantfetchersvc/subscriber.go

```diff
@@ -0,0 +1,122 @@
+	"github.com/kyma-incubator/compass/components/director/internal/labelfilter"
+	"github.com/kyma-incubator/compass/components/director/internal/model"
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/pkg/errors"
+)
+
+// TenantProvisioner is used to create all related to the incoming request tenants, and build their hierarchy;
+//go:generate mockery --name=TenantProvisioner --output=automock --outpkg=automock --case=underscore
+type TenantProvisioner interface {
+	ProvisionTenants(context.Context, *TenantSubscriptionRequest, string) error
+}
+
+// RuntimeService is used to interact with runtimes
+//go:generate mockery --name=RuntimeService --output=automock --outpkg=automock --case=underscore
+type RuntimeService interface {
+	SetLabel(context.Context, *model.LabelInput) error
+	GetLabel(ctx context.Context, runtimeID string, key string) (*model.Label, error)
+	ListByFiltersGlobal(context.Context, []*labelfilter.LabelFilter) ([]*model.Runtime, error)
+}
+
+type subscriptionFunc func(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error
+
+type sliceMutationFunc func([]string, string) []string
+
+type subscriber struct {
+	provisioner TenantProvisioner
+	RuntimeService
+	SubscriptionProviderLabelKey  string
+	ConsumerSubaccountIDsLabelKey string
+}
+
+// NewSubscriber creates new subscriber
+func NewSubscriber(provisioner TenantProvisioner, service RuntimeService, subscriptionProviderLabelKey, consumerSubaccountIDsLabelKey string) *subscriber {
+	return &subscriber{
+		provisioner:                   provisioner,
+		RuntimeService:                service,
+		SubscriptionProviderLabelKey:  subscriptionProviderLabelKey,
+		ConsumerSubaccountIDsLabelKey: consumerSubaccountIDsLabelKey,
+	}
+}
+
+// Subscribe subscribes tenant to runtime. If the tenant does not exist it will be created
+func (s *subscriber) Subscribe(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error {
+	if err := s.provisioner.ProvisionTenants(ctx, tenantSubscriptionRequest, region); err != nil {
+		return err
+	}
+
+	return s.applyRuntimesSubscriptionChange(ctx, tenantSubscriptionRequest.SubscriptionProviderID, tenantSubscriptionRequest.SubaccountTenantID, region, addElement)
+}
+
+// Unsubscribe unsubscribes tenant from runtime.
+func (s *subscriber) Unsubscribe(ctx context.Context, tenantSubscriptionRequest *TenantSubscriptionRequest, region string) error {
+	return s.applyRuntimesSubscriptionChange(ctx, tenantSubscriptionRequest.SubscriptionProviderID, tenantSubscriptionRequest.SubaccountTenantID, region, removeElement)
+}
+
+func (s *subscriber) applyRuntimesSubscriptionChange(ctx context.Context, subscriptionConsumerID, subaccountTenantID, region string, mutateLabelsFunc sliceMutationFunc) error {
+	filters := []*labelfilter.LabelFilter{
+		labelfilter.NewForKeyWithQuery(s.SubscriptionProviderLabelKey, fmt.Sprintf("\"%s\"", subscriptionConsumerID)),
```

Comment:
`subscriptionConsumerID` perhaps this should be `subscriptionProviderID`

Reply:
If I understand correctly we're using _**s.SubscriptionProviderLabelKey**_ to build a filter which will be used to get runtimes labeled with it, and for each runtime to add or remove(depending on the flow - subscribe/unsubscribe) subaccountID label. 
I was wondering who is responsible to set this _**s.SubscriptionProviderLabelKey**_ label because if it missing no runtimes will be returned for processing